File: stone_processor/src/main/java/com/github/klee0kai/stone/helpers/wrap/WrapHelper.java
package com.github.klee0kai.stone.helpers.wrap;

import com.github.klee0kai.stone._hidden_.types.ListUtils;
import com.github.klee0kai.stone._hidden_.types.NullGet;
import com.github.klee0kai.stone.exceptions.StoneException;
import com.github.klee0kai.stone.wrappers.AsyncProvide;
import com.github.klee0kai.stone.wrappers.LazyProvide;
import com.github.klee0kai.stone.wrappers.PhantomProvide;
import com.github.klee0kai.stone.wrappers.Ref;
import com.squareup.javapoet.*;

import javax.inject.Provider;
import java.lang.ref.Reference;
import java.lang.ref.SoftReference;
import java.lang.ref.WeakReference;
import java.util.*;

import static com.github.klee0kai.stone.exceptions.ExceptionStringBuilder.createErrorMes;
import static com.github.klee0kai.stone.utils.ClassNameUtils.noWildCardType;
import static com.github.klee0kai.stone.utils.ClassNameUtils.rawTypeOf;

public class WrapHelper {

    HashMap<TypeName, WrapType> wrapTypes = new HashMap<>();

    public void reInit() {
        wrapTypes.clear();
        std();
    }

    public void support(WrapType wrapType) {
        wrapTypes.putIfAbsent(wrapType.typeName, wrapType);
    }


    public boolean isSupport(TypeName typeName) {
        return wrapTypes.containsKey(rawTypeOf(typeName));
    }

    public boolean isNonCachingWrapper(TypeName typeName) {
        for (TypeName t : allParamTypes(typeName)) {
            WrapType wrapType = wrapTypes.get(rawTypeOf(t));
            if (wrapType != null && wrapType.isNoCachingWrapper)
                return true;
        }
        return false;
    }

    public boolean isAsyncProvider(TypeName typeName) {
        for (TypeName t : allParamTypes(typeName)) {
            WrapType wrapType = wrapTypes.get(rawTypeOf(t));
            if (wrapType != null && wrapType.isAsyncProvider)
                return true;
        }
        return false;
    }

    public boolean isList(TypeName typeName) {
        return ListUtils.indexOf(allParamTypes(typeName), (i, it) -> {
            WrapType wrapType = wrapTypes.get(rawTypeOf(it));
            return wrapType != null && wrapType.isList();
        }) >= 0;
    }


    public TypeName paramType(TypeName typeName) {
        if (typeName instanceof ParameterizedTypeName) {
            ParameterizedTypeName par = (ParameterizedTypeName) typeName;
            if (isSupport(par.rawType) && par.typeArguments != null && !par.typeArguments.isEmpty())
                return par.typeArguments.get(0);
        }
        return typeName;
    }

    /**
     * com.github.klee0kai.stone.wrappers.LazyProvide<com.github.klee0kai.test.tech.components.Battery> -> com.github.klee0kai.test.tech.components.Battery
     * ? extends java.lang.ref.WeakReference<com.github.klee0kai.test.car.model.Window> -> com.github.klee0kai.test.car.model.Window
     */
    public TypeName nonWrappedType(TypeName typeName) {
        if (typeName instanceof ParameterizedTypeName) {
            ParameterizedTypeName par = (ParameterizedTypeName) typeName;
            if (isSupport(par.rawType) && par.typeArguments != null && !par.typeArguments.isEmpty())
                return nonWrappedType(par.typeArguments.get(0));
        }
        if (typeName instanceof WildcardTypeName) {
            WildcardTypeName par = (WildcardTypeName) typeName;
            if (!par.upperBounds.isEmpty())
                return nonWrappedType(par.upperBounds.get(0));
        }
        return typeName;
    }

    /**
     * java.util.List<com.github.klee0kai.stone.wrappers.Ref<com.github.klee0kai.test.boxed.model.CarBox<com.github.klee0kai.test.car.model.Window>>> ->
     * java.util.List<com.github.klee0kai.test.boxed.model.CarBox<com.github.klee0kai.test.car.model.Window>>
     */
    public TypeName listWrapTypeIfNeed(TypeName typeName) {
        if (isList(typeName))
            return ParameterizedTypeName.get(ClassName.get(List.class), nonWrappedType(typeName));
        return nonWrappedType(typeName);
    }

    public List<TypeName> allParamTypes(TypeName typeName) {
        typeName = noWildCardType(typeName);
        List<TypeName> allParams = new LinkedList<>();
        while (true) {
            allParams.add(typeName);
            ParameterizedTypeName paramType = typeName instanceof ParameterizedTypeName ? (ParameterizedTypeName) typeName : null;
            if (paramType == null || paramType.typeArguments.isEmpty()) break;
            typeName = noWildCardType(paramType.typeArguments.get(0));
        }
        return allParams;
    }


    public CodeBlock transform(
            TypeName providingType,
            TypeName wannaType,
            CodeBlock code

    ) {
        if (providingType == null || Objects.equals(providingType, wannaType)) {
            return code;
        }

        CodeBlock.Builder codeBuilder = CodeBlock.builder().add(code);
        LinkedList<TypeName> wrapPathNames = new LinkedList<>(allParamTypes(wannaType));
        LinkedList<TypeName> unwrapPathNames = new LinkedList<>(allParamTypes(providingType));
        Collections.reverse(wrapPathNames);
        while (!wrapPathNames.isEmpty() && !unwrapPathNames.isEmpty()
                && Objects.equals(rawTypeOf(unwrapPathNames.getLast()), rawTypeOf(wrapPathNames.getFirst()))) {
            unwrapPathNames.pollLast();
            wrapPathNames.pollFirst();
        }

        ListUtils.IFormat<TypeName, WrapType> wrapTypeFormat = it -> {
            WrapType type = wrapTypes.get(rawTypeOf(it));
            if (type == null) {
                throw new StoneException(
                        createErrorMes()
                                .typeTransformNonSupport(rawTypeOf(providingType), wannaType)
                                .classNonFound(it.toString())
                                .build(),
                        null
                );
            }
            return type;
        };

        LinkedList<WrapType> unwrapPath = new LinkedList<>(ListUtils.format(unwrapPathNames, wrapTypeFormat));
        LinkedList<WrapType> wrapPath = new LinkedList<>(ListUtils.format(wrapPathNames, wrapTypeFormat));

        while (!unwrapPath.isEmpty()) {
            WrapType unwrapType = unwrapPath.get(0);
            if (unwrapType.isList()) {
                int wrapListIndex = ListUtils.indexOf(wrapPath, (i, it) -> it.isList());
                if (wrapListIndex >= 0) {
                    TypeName unWrapItemType = paramType(unwrapPathNames.get(0));
                    TypeName wrapItemType = paramType(wrapPathNames.get(wrapListIndex));
                    WrapType wrapListType = wrapPath.get(wrapListIndex);

                    codeBuilder = wrapListType.inListFormat.formatCode(
                            unwrapType.typeName,
                            codeBuilder.build(),
                            (listItemCode) ->
                                    transform(
                                            unWrapItemType,
                                            wrapItemType,
                                            listItemCode
                                    )
                    ).toBuilder();

                    for (int i = 0; i <= wrapListIndex; i++) {
                        wrapPath.pollFirst();
                        wrapPathNames.pollFirst();
                    }
                    unwrapPath.clear();
                    unwrapPathNames.clear();
                    break;
                }
            }
            codeBuilder = unwrapType.unwrap.formatCode(codeBuilder.build())
                    .toBuilder();

            unwrapPath.pollFirst();
            unwrapPathNames.pollFirst();
        }

        while (!wrapPath.isEmpty()) {
            codeBuilder = wrapPath.get(0).wrap.formatCode(codeBuilder.build())
                    .toBuilder();

            wrapPath.pollFirst();
            wrapPathNames.pollFirst();
        }

        return codeBuilder.build();
    }

    private void std() {
        for (Class cl : Arrays.asList(WeakReference.class, SoftReference.class, Reference.class)) {
            ClassName wrapper = ClassName.get(cl);
            ClassName creator = !Objects.equals(cl, Reference.class) ? wrapper : ClassName.get(WeakReference.class);

            WrapType wrapType = new WrapType();
            wrapType.isNoCachingWrapper = false;
            wrapType.typeName = wrapper;
            wrapType.wrap = (or) ->
                    CodeBlock.builder()
                            .add("$T.let(", NullGet.class)
                            .add(or)
                            .add(", $T::new )", creator)
                            .build();

            wrapType.unwrap = (or) ->
                    CodeBlock.builder()
                            .add("$T.let( ", NullGet.class)
                            .add(or)
                            .add(", $T::get ) ", cl)
                            .build();

            support(wrapType);
        }

        for (Class cl : Arrays.asList(PhantomProvide.class, Ref.class, Provider.class, LazyProvide.class, AsyncProvide.class)) {
            ClassName wrapper = ClassName.get(cl);

            WrapType wrapType = new WrapType();
            wrapType.isNoCachingWrapper = !Objects.equals(cl, LazyProvide.class) && !Objects.equals(cl, AsyncProvide.class);
            wrapType.isAsyncProvider = true;
            wrapType.typeName = wrapper;

            wrapType.wrap = (or) ->
                    CodeBlock.builder()
                            .add("new $T( () -> ", wrapType.isNoCachingWrapper ? ClassName.get(PhantomProvide.class) : wrapper)
                            .add(or)
                            .add(" )")
                            .build();

            wrapType.unwrap = (or) ->
                    CodeBlock.builder()
                            .add("$T.let( ", NullGet.class)
                            .add(or)
                            .add(", $T::get ) ", cl)
                            .build();
            support(wrapType);
        }

        int index = 0;
        for (Class cl : Arrays.asList(LinkedList.class, ArrayList.class, List.class, Collection.class)) {
            ClassName wrapper = ClassName.get(cl);
            boolean needConstructor = Arrays.asList(LinkedList.class, ArrayList.class).contains(cl);
            ClassName createType = index++ <= 0 ? wrapper : ClassName.get(ArrayList.class);

            WrapType wrapType = new WrapType();
            wrapType.typeName = wrapper;
            wrapType.wrap = (or) -> {
                CodeBlock.Builder builder = CodeBlock.builder();
                if (needConstructor) builder.add("$T.let( ", NullGet.class);
                builder.add("$T.list( $L ) ", NullGet.class, or);
                if (needConstructor) builder.add(", $T::new)", createType);
                return builder.build();
            };

            wrapType.unwrap = (or) -> CodeBlock.builder()
                    .add("$T.first( $L )", ListUtils.class, or)
                    .build();

            wrapType.inListFormat = (originalListType, originalListCode, itemTransformFun) -> {
                CodeBlock.Builder builder = CodeBlock.builder();
                boolean isListNeedConstructor = needConstructor
                        || originalListType != null && !Objects.equals(rawTypeOf(wrapper), rawTypeOf(originalListType));

                if (isListNeedConstructor) builder.add("$T.let( ", NullGet.class);

                CodeBlock itemTransform = itemTransformFun.formatCode(CodeBlock.of("it"));
                if (Objects.equals(itemTransform.toString(), "it")) {
                    //no transforms
                    builder.add(originalListCode);
                } else {
                    builder.add("$T.format( ", ListUtils.class)
                            .add(originalListCode)
                            .add(", it ->  ")
                            .add(itemTransformFun.formatCode(CodeBlock.of("it")))
                            .add(") ");

                }
                if (isListNeedConstructor) builder.add(", $T::new)", createType);

                return builder.build();
            };
            support(wrapType);

        }
    }


}

File: stone_ksp/src/commonMain/kotlin/com/github/klee0kai/thekey/stone/ksp/helpers/wrap/WrapHelper.kt
package com.github.klee0kai.thekey.stone.ksp.helpers.wrap

import com.github.klee0kai.stone.Provider
import com.github.klee0kai.stone.weakref.Ref
import com.github.klee0kai.stone.weakref.SoftRef
import com.github.klee0kai.stone.weakref.WeakRef
import com.github.klee0kai.stone.wrappers.AsyncLazy
import com.github.klee0kai.stone.wrappers.AsyncProvider
import com.github.klee0kai.stone.wrappers.LazyProvider
import com.github.klee0kai.thekey.stone.ksp.exceptions.StoneException
import com.github.klee0kai.thekey.stone.ksp.helpers.wrap.ClassNameUtils.noWildCardType
import com.github.klee0kai.thekey.stone.ksp.helpers.wrap.ClassNameUtils.rawTypeOf
import com.github.klee0kai.thekey.stone.ksp.poet.codeBlock
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import java.lang.ref.Reference
import java.lang.ref.SoftReference
import java.lang.ref.WeakReference
import java.util.*


class WrapHelper {

    private var wrapTypes = HashMap<TypeName, WrapType>()

    init {
        std()
    }

    fun support(wrapType: WrapType) {
        wrapTypes.putIfAbsent(wrapType.typeName, wrapType)
    }

    fun isSupport(typeName: TypeName): Boolean = wrapTypes.containsKey(rawTypeOf(typeName))

    fun isNonCachingWrapper(typeName: TypeName): Boolean {
        for (t in allParamTypes(typeName)) {
            val wrapType = wrapTypes.get(rawTypeOf(t))
            if (wrapType != null && wrapType.isNoCachingWrapper) return true
        }
        return false
    }

    fun isList(typeName: TypeName): Boolean = allParamTypes(typeName).any {
        val wrapType = wrapTypes.get(rawTypeOf(it))
        wrapType != null && wrapType.isList
    }

    fun paramType(typeName: TypeName): TypeName {
        if (typeName is ParameterizedTypeName) {
            if (isSupport(typeName.rawType) && !typeName.typeArguments.isEmpty()) {
                return typeName.typeArguments.first()
            }
        }
        return typeName
    }

    /**
     * com.github.klee0kai.stone.wrappers.LazyProvide<com.github.klee0kai.test.tech.components.Battery> -> com.github.klee0kai.test.tech.components.Battery
     * ? extends java.lang.ref.WeakReference<com.github.klee0kai.test.car.model.Window> -> com.github.klee0kai.test.car.model.Window
    </com.github.klee0kai.test.car.model.Window></com.github.klee0kai.test.tech.components.Battery> */
    fun nonWrappedType(typeName: TypeName): TypeName {
        if (typeName is ParameterizedTypeName) {
            if (isSupport(typeName.rawType) && !typeName.typeArguments.isEmpty()) {
                return nonWrappedType(typeName.typeArguments.first())
            }
        }
        if (typeName is WildcardTypeName) {
            if (!typeName.outTypes.isEmpty()) return nonWrappedType(typeName.outTypes.first())
        }
        return typeName.copy(nullable = false)
    }

    /**
     * java.util.List<com.github.klee0kai.stone.wrappers.Ref></com.github.klee0kai.stone.wrappers.Ref><com.github.klee0kai.test.boxed.model.CarBox></com.github.klee0kai.test.boxed.model.CarBox><com.github.klee0kai.test.car.model.Window>>> ->
     * java.util.List<com.github.klee0kai.test.boxed.model.CarBox></com.github.klee0kai.test.boxed.model.CarBox><com.github.klee0kai.test.car.model.Window>>
    </com.github.klee0kai.test.car.model.Window></com.github.klee0kai.test.car.model.Window> */
    fun listWrapTypeIfNeed(typeName: TypeName): TypeName {
        if (isList(typeName)) return List::class.asClassName().parameterizedBy(nonWrappedType(typeName))
        return nonWrappedType(typeName)
    }

    fun allParamTypes(typeName: TypeName): List<TypeName> {
        var typeName = noWildCardType(typeName)
        val allParams = LinkedList<TypeName>()
        while (true) {
            allParams.add(typeName)
            val paramType = typeName as? ParameterizedTypeName
            if (paramType == null || paramType.typeArguments.isEmpty()) break
            typeName = noWildCardType(paramType.typeArguments[0])
        }
        return allParams
    }


    fun transform(
        providingType: TypeName,
        wannaType: TypeName,
        code: CodeBlock
    ): CodeBlock {
        if (providingType == wannaType) return code

        var codeBuilder = CodeBlock.builder().add(code)
        val wrapPathNames = LinkedList(allParamTypes(wannaType))
        val unwrapPathNames = LinkedList(allParamTypes(providingType))
        wrapPathNames.reverse()

        while (!wrapPathNames.isEmpty() && !unwrapPathNames.isEmpty()
            && unwrapPathNames.last().rawType() == wrapPathNames.first().rawType()
        ) {
            unwrapPathNames.pollLast()
            wrapPathNames.pollFirst()
        }

        val wrapTypeFormat: (TypeName) -> WrapType = { it: TypeName ->
            wrapTypes[it.rawType()]
                ?: throw StoneException(message = "Type Transform non support $providingType -> $wannaType")
        }

        val unwrapPath = LinkedList(unwrapPathNames.map { wrapTypeFormat(it).copy(typeName = it) })
        val wrapPath = LinkedList(wrapPathNames.map { wrapTypeFormat(it).copy(typeName = it) })

        var currentNullable = providingType.isNullable

        while (!unwrapPath.isEmpty()) {
            val unwrapType = unwrapPath.first()
            if (unwrapType.isList) {
                val wrapListIndex = wrapPath.indexOfFirst { it.isList }
                if (wrapListIndex >= 0) {
                    val unWrapItemType = paramType(unwrapPathNames[0])
                    val wrapItemType = paramType(wrapPathNames[wrapListIndex])
                    val wrapListType = wrapPath[wrapListIndex]

                    codeBuilder = wrapListType.inListFormat!!.formatCode(
                        originalListType = unwrapType.typeName,
                        or = codeBuilder.build(),
                        itemTransformFun = { listItemCode ->
                            transform(unWrapItemType, wrapItemType, listItemCode)
                        }
                    ).toBuilder()

                    for (i in 0..wrapListIndex) {
                        wrapPath.pollFirst()
                        wrapPathNames.pollFirst()
                    }
                    unwrapPath.clear()
                    unwrapPathNames.clear()
                    break
                }
            }
            unwrapPath.pollFirst()
            unwrapPathNames.pollFirst()

            currentNullable = (
                    wrapPath.firstOrNull()?.typeName?.let { paramType(it) }
                        ?: wannaType)
                .isNullable

            codeBuilder = unwrapType.unwrap.formatCode(
                or = codeBuilder.build(),
                srcNullable = unwrapType.typeName.isNullable,
                targetNullable = currentNullable
            ).toBuilder()
        }

        while (!wrapPath.isEmpty()) {
            val wrapType = wrapPath.first()

            wrapPath.pollFirst()
            wrapPathNames.pollFirst()

            codeBuilder = wrapType.wrap.formatCode(
                codeBuilder.build(),
                srcNullable = currentNullable,
                targetNullable = wrapType.typeName.isNullable,
                targetArgTypeNullable = paramType(wrapType.typeName).isNullable,
            ).toBuilder()

            currentNullable = wrapType.typeName.isNullable
        }

        if (currentNullable && !wannaType.isNullable) {
            codeBuilder.add("!!")
        }

        return codeBuilder.build()
    }

    private fun std() {
        for (cl in listOf(
            WeakReference::class,
            SoftReference::class,
            Reference::class,
        )) {
            val wrapper = cl.asClassName()
            val creator = if (cl != Reference::class) wrapper else WeakReference::class.asClassName()

            val wrapType = WrapType(
                typeName = wrapper,
                isNoCachingWrapper = false,
                wrap = { or, srcNullable, targetNullable, argTypeNullable ->
                    codeBlock {
                        when {
                            !srcNullable -> add("%T( %L )", creator, or)
                            targetNullable -> add("%L?.let{ %T( it ) }", or, creator)
                            else -> add("%T( %L!! )", creator, or)
                        }
                    }
                },
                unwrap = { or, srcNullable, targetNullable ->
                    codeBlock {
                        when {
                            targetNullable -> add("%L?.get()", or)
                            else -> add("%L!!.get()!!", or)
                        }
                    }
                }
            )
            support(wrapType)
        }


        for (cl in listOf(
            WeakRef::class,
            SoftRef::class,
        )) {
            val creator = cl.asClassName()

            val wrapType = WrapType(
                typeName = creator,
                isNoCachingWrapper = false,
                wrap = { or, srcNullable, targetNullable, argTypeNullable ->
                    codeBlock {
                        when {
                            !srcNullable -> add("%T( %L )", creator, or)
                            targetNullable -> add("%L?.let{ %T( it ) }", or, creator)
                            else -> add("%T( %L!! )", creator, or)
                        }
                    }
                },
                unwrap = { or, srcNullable, targetNullable ->
                    codeBlock {
                        when {
                            targetNullable -> add("%L?.get()", or)
                            else -> add("%L!!.get()!!", or)
                        }
                    }
                }
            )
            support(wrapType)
        }

        for (cl in listOf(
            Lazy::class,
        )) {
            val wrapper = cl.asClassName()
            val wrapType = WrapType(
                typeName = wrapper,
                isNoCachingWrapper = false,
                wrap = { or, srcNullable, targetNullable, argTypeNullable ->
                    codeBlock {
                        when {
                            !srcNullable || argTypeNullable -> add("lazy{ %L } ", or)
                            else -> add("lazy{ %L!! } ", or)
                        }
                    }
                },
                unwrap = { or, srcNullable, targetNullable ->
                    codeBlock {
                        when {
                            targetNullable -> add("%L?.value", or)
                            !srcNullable -> add("%L.value", or)
                            else -> add("%L!!.value", or)
                        }
                    }
                },
            )
            support(wrapType)
        }

        for (cl in listOf(
            Provider::class,
            Ref::class,
            javax.inject.Provider::class,
            LazyProvider::class,
            AsyncLazy::class,
            AsyncProvider::class,
        )) {
            val isNoCachingWrapper = cl != LazyProvider::class && cl != AsyncLazy::class

            val wrapper = cl.asClassName()
            val wrapType = WrapType(
                typeName = wrapper,
                isNoCachingWrapper = isNoCachingWrapper,
                wrap = { or, srcNullable, targetNullable, argTypeNullable ->
                    codeBlock {
                        when {
                            !srcNullable || argTypeNullable -> add("%T{ %L }", wrapper, or)
                            else -> add("%T{ %L!! } ", wrapper, or)
                        }
                    }
                },
                unwrap = { or, srcNullable, targetNullable ->
                    codeBlock {
                        when {
                            targetNullable -> add("%L?.get()", or)
                            !srcNullable -> add("%L.get()", or)
                            else -> add("%L!!.get()!!", or)
                        }
                    }
                },
            )
            support(wrapType)
        }



        for (cl in listOf(
            LinkedList::class,
            ArrayList::class,
            List::class,
            Collection::class,
        )) {
            val wrapper = cl.asClassName()
            val constructor = when (cl) {
                LinkedList::class, ArrayList::class -> wrapper
                else -> null
            }

            val wrapType = WrapType(
                typeName = wrapper,
                isNoCachingWrapper = false,
                wrap = { or, srcNullable, targetNullable, argTypeNullable ->
                    codeBlock {
                        when {
                            constructor != null && targetNullable -> add("%L?.let { %T(it) }", or, constructor)
                            targetNullable -> add("%L?.let { listOfNotNull( it ) }", or)
                            constructor != null && argTypeNullable -> add("%L!!.let { %T(it) }", or, constructor)
                            constructor != null -> add("%L!!.let { %T(it!!) }", or, constructor)
                            else -> add("listOfNotNull( %L )", or)
                        }

                    }
                },
                unwrap = { or, srcNullable, targetNullable ->
                    codeBlock {
                        when {
                            targetNullable -> add("%L?.firstOrNull()", or)
                            !srcNullable -> add("%L.first()", or)
                            else -> add("%L!!.first()", or)
                        }
                    }
                },
                inListFormat = { originalListType, originalListCode, itemTransformFun ->
                    codeBlock {
                        val itemTransform = itemTransformFun.formatCode(CodeBlock.of("it"))
                        if (itemTransform.toString() == "it") {
                            //no transforms
                            add(originalListCode)
                        } else {
                            add("( %L?.map{ it -> %L } ?: emptyList() )", originalListCode, itemTransform)
                        }

                        if (wrapper.rawType() != originalListType.rawType()) {
                            if (constructor != null) {
                                add("!!.let { %T(it) }", constructor)
                            } else {
                                add("!!.toList()")
                            }
                        }
                    }
                },
            )

            support(wrapType)
        }
    }
}